File: examples/runtimes/java/src/main/java/com/blok/runtime/Blok.java
package com.blok.runtime;

import java.util.Map;

/**
 * Blok SDK for Java - Core types and interfaces
 */
public class Blok {

    /**
     * Context represents the workflow execution context
     */
    public static class Context {
        public String id;
        public String workflow_name;
        public String workflow_path;
        public Request request;
        public Response response;
        public Map<String, Object> vars;
        public Map<String, String> env;

        public Context() {}
    }

    /**
     * Request represents the incoming HTTP request data
     */
    public static class Request {
        public Object body;
        public Map<String, String> headers;
        public Map<String, String> params;
        public Map<String, String> query;
        public String method;
        public String url;
        public Map<String, String> cookies;
        public String baseUrl;

        public Request() {}
    }

    /**
     * Response represents the workflow response
     */
    public static class Response {
        public Object data;
        public String contentType;
        public boolean success;
        public Object error;

        public Response() {
            this.success = true;
        }
    }

    /**
     * NodeConfig represents node-specific configuration
     */
    public static class NodeConfig {
        public String name;
        public String path;
        public Map<String, Object> config;

        public NodeConfig() {}
    }

    /**
     * ExecutionRequest is the request received from the Blok runner
     */
    public static class ExecutionRequest {
        public NodeConfig node;
        public Context context;

        public ExecutionRequest() {}
    }

    /**
     * ExecutionResult is the response returned to the Blok runner
     */
    public static class ExecutionResult {
        public boolean success;
        public Object data;
        public Object errors;
        public String[] logs;
        public Map<String, Object> metrics;

        public ExecutionResult() {
            this.success = true;
        }

        public ExecutionResult(boolean success, Object data, Object errors) {
            this.success = success;
            this.data = data;
            this.errors = errors;
        }
    }

    /**
     * NodeHandler is the interface that all Blok nodes must implement
     */
    public interface NodeHandler {
        Object execute(Context context, Map<String, Object> config) throws Exception;
    }

    /**
     * HealthStatus represents the health status of the runtime
     */
    public static class HealthStatus {
        public String status;
        public String version;
        public String[] nodes_loaded;

        public HealthStatus(String version, String[] nodesLoaded) {
            this.status = "healthy";
            this.version = version;
            this.nodes_loaded = nodesLoaded;
        }
    }
}

File: examples/runtimes/go/sdk/blok.go
// Package blok provides the Blok runtime SDK for Go
package blok

import (
	"encoding/json"
	"fmt"
)

// Context represents the workflow execution context
type Context struct {
	ID           string                 `json:"id"`
	WorkflowName string                 `json:"workflow_name"`
	WorkflowPath string                 `json:"workflow_path"`
	Request      Request                `json:"request"`
	Response     Response               `json:"response"`
	Vars         map[string]interface{} `json:"vars"`
	Env          map[string]string      `json:"env"`
}

// Request represents the incoming HTTP request data
type Request struct {
	Body    interface{}       `json:"body"`
	Headers map[string]string `json:"headers"`
	Params  map[string]string `json:"params"`
	Query   map[string]string `json:"query"`
	Method  string            `json:"method"`
	URL     string            `json:"url"`
	Cookies map[string]string `json:"cookies"`
	BaseURL string            `json:"baseUrl"`
}

// Response represents the workflow response
type Response struct {
	Data        interface{} `json:"data"`
	ContentType string      `json:"contentType"`
	Success     bool        `json:"success"`
	Error       interface{} `json:"error"`
}

// NodeConfig represents node-specific configuration
type NodeConfig struct {
	Name   string                 `json:"name"`
	Path   string                 `json:"path"`
	Config map[string]interface{} `json:"config"`
}

// ExecutionRequest is the request received from the Blok runner
type ExecutionRequest struct {
	Node    NodeConfig `json:"node"`
	Context Context    `json:"context"`
}

// ExecutionResult is the response returned to the Blok runner
type ExecutionResult struct {
	Success bool                   `json:"success"`
	Data    interface{}            `json:"data"`
	Errors  interface{}            `json:"errors"`
	Logs    []string               `json:"logs,omitempty"`
	Metrics map[string]interface{} `json:"metrics,omitempty"`
}

// NodeHandler is the interface that all Blok nodes must implement
type NodeHandler interface {
	Execute(ctx *Context, config map[string]interface{}) (interface{}, error)
}

// NodeRegistry manages registered node handlers
type NodeRegistry struct {
	nodes map[string]NodeHandler
}

// NewNodeRegistry creates a new node registry
func NewNodeRegistry() *NodeRegistry {
	return &NodeRegistry{
		nodes: make(map[string]NodeHandler),
	}
}

// Register registers a node handler with the given name
func (r *NodeRegistry) Register(name string, handler NodeHandler) {
	r.nodes[name] = handler
}

// Get retrieves a node handler by name
func (r *NodeRegistry) Get(name string) (NodeHandler, error) {
	handler, exists := r.nodes[name]
	if !exists {
		return nil, fmt.Errorf("node '%s' not found", name)
	}
	return handler, nil
}

// Execute executes a node by name
func (r *NodeRegistry) Execute(req *ExecutionRequest) *ExecutionResult {
	handler, err := r.Get(req.Node.Name)
	if err != nil {
		return &ExecutionResult{
			Success: false,
			Data:    nil,
			Errors: map[string]string{
				"message": err.Error(),
			},
		}
	}

	data, err := handler.Execute(&req.Context, req.Node.Config)
	if err != nil {
		return &ExecutionResult{
			Success: false,
			Data:    nil,
			Errors: map[string]string{
				"message": err.Error(),
			},
		}
	}

	return &ExecutionResult{
		Success: true,
		Data:    data,
		Errors:  nil,
	}
}

// HealthStatus represents the health status of the runtime
type HealthStatus struct {
	Status      string   `json:"status"`
	Version     string   `json:"version"`
	NodesLoaded []string `json:"nodes_loaded"`
}

// GetHealth returns the health status
func (r *NodeRegistry) GetHealth(version string) *HealthStatus {
	nodes := make([]string, 0, len(r.nodes))
	for name := range r.nodes {
		nodes = append(nodes, name)
	}

	return &HealthStatus{
		Status:      "healthy",
		Version:     version,
		NodesLoaded: nodes,
	}
}

// Helper function to parse JSON body into a struct
func ParseBody(body interface{}, target interface{}) error {
	bytes, err := json.Marshal(body)
	if err != nil {
		return err
	}
	return json.Unmarshal(bytes, target)
}
